internal/authz/authz.go

```diff
@@ -0,0 +1,434 @@
+	Authorize(ctx context.Context, storeID, apiMethod string, modules ...string) error
+	AuthorizeCreateStore(ctx context.Context) error
+	AuthorizeListStores(ctx context.Context) error
+	ListAuthorizedStores(ctx context.Context) ([]string, error)
+	GetModulesForWriteRequest(req *openfgav1.WriteRequest, typesys *typesystem.TypeSystem) ([]string, error)
+}
+
+type NoopAuthorizer struct {
+	config *Config
+	server ServerInterface
+	logger logger.Logger
+}
+
+func NewAuthorizerNoop(config *Config, server ServerInterface, logger logger.Logger) *NoopAuthorizer {
+	return &NoopAuthorizer{
+		config: config,
+		server: server,
+		logger: logger,
+	}
+}
+
+func (a *NoopAuthorizer) Authorize(ctx context.Context, storeID, apiMethod string, modules ...string) error {
+	return nil
+}
+
+func (a *NoopAuthorizer) AuthorizeCreateStore(ctx context.Context) error {
+	return nil
+}
+
+func (a *NoopAuthorizer) AuthorizeListStores(ctx context.Context) error {
+	return nil
+}
+
+func (a *NoopAuthorizer) ListAuthorizedStores(ctx context.Context) ([]string, error) {
+	return nil, nil
+}
+
+func (a *NoopAuthorizer) GetModulesForWriteRequest(req *openfgav1.WriteRequest, typesys *typesystem.TypeSystem) ([]string, error) {
+	return nil, nil
+}
+
+type Authorizer struct {
+	config *Config
+	server ServerInterface
+	logger logger.Logger
+}
+
+func NewAuthorizer(config *Config, server ServerInterface, logger logger.Logger) *Authorizer {
+	return &Authorizer{
+		config: config,
+		server: server,
+		logger: logger,
+	}
+}
+
+type AuthorizationError struct {
+	Err error
+}
+
+func (e *AuthorizationError) Error() string {
+	return fmt.Sprintf("error authorizing request: %s", e.Err)
+}
+
+func (e *AuthorizationError) Unwrap() error {
+	return e.Err
+}
+
+func (a *Authorizer) getRelation(apiMethod string) (string, error) {
+	switch apiMethod {
+	case ReadAuthorizationModel, ReadAuthorizationModels:
+		return CanCallReadAuthorizationModels, nil
+	case Read:
+		return CanCallRead, nil
+	case Write:
+		return CanCallWrite, nil
+	case ListObjects, StreamedListObjects:
+		return CanCallListObjects, nil
+	case Check:
+		return CanCallCheck, nil
+	case ListUsers:
+		return CanCallListUsers, nil
+	case WriteAssertions:
+		return CanCallWriteAssertions, nil
+	case ReadAssertions:
+		return CanCallReadAssertions, nil
+	case WriteAuthorizationModel:
+		return CanCallWriteAuthorizationModels, nil
+	case ListStores:
+		return CanCallListStores, nil
+	case CreateStore:
+		return CanCallCreateStore, nil
+	case GetStore:
+		return CanCallGetStore, nil
+	case DeleteStore:
+		return CanCallDeleteStore, nil
+	case Expand:
+		return CanCallExpand, nil
+	case ReadChanges:
+		return CanCallReadChanges, nil
+	default:
+		return "", AuthorizationError{Err: ErrUnknownAPIMethod}.Err
+	}
+}
+
+// Authorize checks if the user has access to the resource.
+func (a *Authorizer) Authorize(ctx context.Context, storeID, apiMethod string, modules ...string) error {
+	claims, err := checkAuthClaims(ctx)
```

Comment:
```suggestion
var tracer = otel.Tracer("internal/authz")

func (a *Authorizer) Authorize(ctx context.Context, storeID, apiMethod string, modules ...string) error {
       ctx, span := tracer.Start(ctx, "Authorize", trace.WithAttributes(
		attribute.String("storeID", storeID),
		attribute.String("apiMethod", apiMethod),
		attribute.String("modules", strings.Join(modules, ",")),
	))
	defer span.End()
	claims, err := checkAuthClaims(ctx)
```

This creates a new span in the trace. It's a good idea to do this in every function where something important happens. It helps me (and i'm sure others too) debug issues

<img src="https://github.com/user-attachments/assets/5c1714b4-1bc1-44c5-af25-a6a5eadd35b4">